File: src/core/resources.ts
// Port of Course/basicresources.{h,cpp} and Core/resourcemaps.h.
// All numbers are copied verbatim from the original resourcemaps.h.

export enum BasicResource {
  NONE = 0,
  MONEY = 1,
  WOOD = 2,
  STONE = 3,
  METAL = 4,
}

/**
 * ResourceMap mirrors std::map<BasicResource, int>. We use a Map so that
 * iteration semantics (only present keys are visited) match the C++ original,
 * which several algorithms rely on (mergeResourceMaps, hasEnoughResources).
 */
export type ResourceMap = Map<BasicResource, number>;

export function rmap(entries: Partial<Record<BasicResource, number>>): ResourceMap {
  const m: ResourceMap = new Map();
  for (const [k, v] of Object.entries(entries)) {
    m.set(Number(k) as BasicResource, v as number);
  }
  return m;
}

/** Deep copy. */
export function cloneResourceMap(map: ResourceMap): ResourceMap {
  return new Map(map);
}

/**
 * mergeResourceMaps: start from `right`, then for every key in `left` either
 * insert it (if absent) or add to the existing value. Keys present in only one
 * map survive. (Course::mergeResourceMaps)
 */
export function mergeResourceMaps(left: ResourceMap, right: ResourceMap): ResourceMap {
  const result = new Map(right);
  for (const [key, value] of left) {
    if (!result.has(key)) {
      result.set(key, value);
    } else {
      result.set(key, result.get(key)! + value);
    }
  }
  return result;
}

/** reverseResourceMap: negate every value. */
export function reverseResourceMap(map: ResourceMap): ResourceMap {
  const result = new Map<BasicResource, number>();
  for (const [key, value] of map) result.set(key, value * -1);
  return result;
}

/** getNegativesMap: keep negatives, zero out the rest. */
export function getNegativesMap(map: ResourceMap): ResourceMap {
  const result = new Map<BasicResource, number>();
  for (const [key, value] of map) result.set(key, value < 0 ? value : 0);
  return result;
}

/** getPositivesMap: keep >=0 values, zero out negatives. */
export function getPositivesMap(map: ResourceMap): ResourceMap {
  const result = new Map<BasicResource, number>();
  for (const [key, value] of map) result.set(key, value >= 0 ? value : 0);
  return result;
}

// ---------------------------------------------------------------------------
// ConstResourceMaps (verbatim from resourcemaps.h)
// ---------------------------------------------------------------------------

export const EMPTY: ResourceMap = rmap({});

export const NO_RESOURCES: ResourceMap = rmap({
  [BasicResource.MONEY]: 0,
  [BasicResource.WOOD]: 0,
  [BasicResource.METAL]: 0,
  [BasicResource.STONE]: 0,
});

export const RESOURCE_LIMITS: ResourceMap = rmap({
  [BasicResource.MONEY]: 9999999,
  [BasicResource.WOOD]: 9999999,
  [BasicResource.STONE]: 9999999,
  [BasicResource.METAL]: 9999999,
});

export const UNIT_LIMITS = 999;

export const STARTING_RESOURCES: ResourceMap = rmap({
  [BasicResource.MONEY]: 400,
  [BasicResource.WOOD]: 200,
  [BasicResource.STONE]: 100,
  [BasicResource.METAL]: 25,
});

// Tile - Forest
export const FOREST_PRODUCTION: ResourceMap = rmap({
  [BasicResource.WOOD]: 100,
  [BasicResource.STONE]: 10,
});
export const FOREST_CAPACITY: ResourceMap = rmap({
  [BasicResource.WOOD]: 600,
  [BasicResource.STONE]: 60,
});

// Tile - Abundant Forest
export const ABUNDANT_FOREST_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: 15,
});

export const FOREST_GROW_TIME = 5;

// Building - Farm
export const FARM_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -100,
  [BasicResource.WOOD]: -100,
  [BasicResource.METAL]: -5,
});
export const FARM_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: 175,
});
export const FARM_GROW_TIME = 4;

// Building - Mine
// NOTE: economy values below DIVERGE from the original C++ resourcemaps.h (a deliberate
// rebalance to make industry a real choice, not a 1:1 port value). See CLAUDE.md.
// Mine is the material engine (metal funds the army/outposts, stone funds villages/plants);
// its build cost was lowered so that engine comes online sooner. Production unchanged so
// the military economy (metal supply) keeps its balance.
export const MINE_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -200,
  [BasicResource.WOOD]: -200,
  [BasicResource.STONE]: 200,
});
export const MINE_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: 20,
  [BasicResource.STONE]: 30,
  [BasicResource.METAL]: 20,
});

// Building - Hydroelectric Power Plant
// Rebalanced (diverges from C++): cheaper and far more productive (40 -> 80/worker) so that
// claiming a straight-river tile is genuinely worth it — a mid-game money building that
// beats a farm per unit-slot while sitting on otherwise low-value river land.
export const HEPP_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -280,
  [BasicResource.WOOD]: -150,
  [BasicResource.STONE]: -120,
  [BasicResource.METAL]: -60,
});
export const HEPP_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: 80,
});

// Building - Nuclear Power Plant
// Rebalanced (diverges from C++): the late-game economic engine. The cost is shifted into
// MONEY (1200 -> 2000) and away from materials (500/500 -> 250/250) so it is a long money
// *savings goal* rather than a mine-grind, and the payoff is big — 100 -> 160/worker, so an
// expert + 2 workers nets ~285/round (~2.4x a farm per unit-slot). Save up, then dominate.
export const NUCLEARPP_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -2000,
  [BasicResource.WOOD]: -200,
  [BasicResource.STONE]: -250,
  [BasicResource.METAL]: -250,
});
export const NUCLEARPP_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: 160,
});

// Building - Outpost (REBALANCED — diverges from C++; this is the source of truth)
// The C++/original cost (650 money + 300 wood + 300 stone + 300 METAL) made the Outpost —
// and therefore the entire soldier-cap → army chain — UNREACHABLE on a normal economy: 300
// metal at once needs ~15 mine-rounds of pure hoarding (a mine yields 20 metal/round) and a
// normal game runs ~1 mine, so the candidate was never enumerated and the soldier cap stayed
// hard-locked at 1 (HQ +1). Same "always-worse, never-chosen" trap the Mine/Hydro/Nuclear
// tier was rebalanced out of (see CLAUDE.md). New cost: metal 300→100 (≈5 mine-rounds, ~3
// with an expert mine), money 650→500, wood/stone 300→200 — the army stays a deliberate
// multi-resource commitment but is now a REACHABLE real choice mid-game. Stays in lockstep
// with outpost_build_cost() in rust-trainer/crates/cp-sim/src/resources.rs.
export const OUTPOST_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -500,
  [BasicResource.WOOD]: -200,
  [BasicResource.STONE]: -200,
  [BasicResource.METAL]: -100,
});
// Per-round Outpost upkeep. METAL upkeep REBALANCED -15 → -5 (arc sd2 → sd3, 2026-06-06,
// "military-economy" pass). The -15 metal/round drain meant a single mine (≈20 metal/round)
// could barely carry ONE Outpost before going metal-negative, so the soldier-cap → army
// chain stalled at 1-2 soldiers and military was never fundable (see CLAUDE.md). At -5/round
// ~1 mine comfortably carries 2-3 Outposts (cap 7-10 soldiers). The -50 money upkeep, the
// outpost build cost, mine output and the Device soldier-cap halving are all left UNCHANGED
// so a pure-economy line stays competitive. Mirrors outpost_production() in
// rust-trainer/crates/cp-sim/src/resources.rs (parity-locked).
export const OUTPOST_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: -50,
  [BasicResource.METAL]: -5,
});
export const OUTPOST_SOLDIER_VALUE = 3;

// Building - Bridge
export const BRIDGE_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -100,
  [BasicResource.WOOD]: -300,
  [BasicResource.STONE]: -150,
});
export const BRIDGE_PRODUCTION: ResourceMap = rmap({
  [BasicResource.WOOD]: -5,
});

// Building - Village (Neighborhood)
// MONEY upkeep REBALANCED -10 → -5 (arc sd3 → sd4, 2026-06-08, "unit-cap economy" pass).
// Unit cap = HQ(+3) + Σ Village(+3) + Mikontalo(+2); the ONLY scalable source of cap is
// the Village, at -10 money/round each. Funding BOTH a ~3-4 soldier army AND winning
// expansion needs cap to staff mines (2 workers + expert = 80 metal) + farms + experts
// simultaneously — typically 3 Villages beyond HQ, i.e. -30 money/round of pure cap
// overhead at -10. That overhead made army and winning expansion a genuine tradeoff
// (every PPO/AZ champion capped ~0.62-0.66 trueWin; see CLAUDE.md "unit-cap economy
// rebalance"). At -5 the same 3-Village cap costs -15/round — ~one staffed farm's net
// income (≈+39) restored toward expansion, enough to carry the army's Outpost upkeep
// (-50) + a couple soldier salaries while still expanding. The -10 wood/-10 stone upkeep,
// the build cost, and VILLAGE_UNIT_VALUE are left UNCHANGED so a Village stays a real
// commitment and the cap-per-Village stays put; only the binding money drain is halved.
// Mirrors village_production() in rust-trainer/crates/cp-sim/src/resources.rs
// (parity-locked) and the village net-money model in metrics.ts/metrics.rs + the
// buildVillage netDelta in candidates.ts/candidates.rs.
export const VILLAGE_BUILD_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -200,
  [BasicResource.WOOD]: -200,
  [BasicResource.STONE]: -100,
  [BasicResource.METAL]: -25,
});
export const VILLAGE_PRODUCTION: ResourceMap = rmap({
  [BasicResource.MONEY]: -5,
  [BasicResource.WOOD]: -10,
  [BasicResource.STONE]: -10,
});
export const VILLAGE_UNIT_VALUE = 3;

// Building - Strange Device
// DELIBERATE divergence from the C++/Qt original (the original had no Device) — a new
// decisive, draw-eliminating win condition. See STRANGE-DEVICE-DESIGN.md and CLAUDE.md.
// The build cost is a one-time nuclear-tier commitment with NO per-turn drain; the
// balancer is the soldier-cap halving (PlayerBase.updateUnitAmounts), not the economy.
// These numbers are TUNABLE — set empirically via the sim/ draw-rate + builder-win-rate
// measurement (target: timeout% → 0 AND builder-vs-non-builder win% ≈ 50%).
export const STRANGE_DEVICE_BUILD_COST: ResourceMap = rmap({
  // Tuned 2026-06-03 to 1300 (from a 1800 first pass): at 1800 the halved-army builder
  // was 56% of all bankruptcies and couldn't afford to DEFEND its own Device; 1300 keeps
  // it a real commitment while dropping bankruptcies (~27%->~21% in hard-vs-hard), making
  // the Device more central (~39% of games end by it) and Device-survival ~coin-flip
  // (~53%). Kept in lockstep with cp-sim strange_device_build_cost().
  [BasicResource.MONEY]: -1300,
  [BasicResource.STONE]: -200,
  [BasicResource.METAL]: -200,
});
// Countdown (in the owner's end-of-turns ≈ rounds) before a standing Device wins.
// Scales with map size: on a bigger map it takes longer to mass an army and cross the
// map to crack the Device, so the countdown must be longer to stay a genuine threat
// rather than an instant win. round(BASE + PER_TILE × tileCount): 10x10→22, 12x12→26,
// 14x12→29, 16x16→38. Trimmed for arc sd5 (was 18 + 0.12) so the device can land before
// a conquest race closes the game — a real alternate win path, not always-too-slow.
// Tunable.
export const STRANGE_DEVICE_COUNTDOWN_BASE = 12;
export const STRANGE_DEVICE_COUNTDOWN_PER_TILE = 0.1;
export function strangeDeviceCountdown(tileCount: number): number {
  return Math.round(STRANGE_DEVICE_COUNTDOWN_BASE + STRANGE_DEVICE_COUNTDOWN_PER_TILE * tileCount);
}

// Building - Mikontalo
export const MIKONTALO_UNIT_VALUE = 2;

// Building - HQ
export const HQ_UNIT_VALUE = 3;
export const HQ_SOLDIER_VALUE = 1;

// Worker
export const BASIC_WORKER_COST: ResourceMap = rmap({ [BasicResource.MONEY]: -50 });
export const BASIC_WORKER_SALARY: ResourceMap = rmap({ [BasicResource.MONEY]: -5 });

// Expert
export const EXPERT_COST: ResourceMap = rmap({ [BasicResource.MONEY]: -250 });
export const EXPERT_SALARY: ResourceMap = rmap({ [BasicResource.MONEY]: -25 });

// Soldier
// METAL build cost REBALANCED -50 → -30 (arc sd2 → sd3, 2026-06-06, "military-economy" pass).
// Paired with the Outpost metal-upkeep cut above so a single-mine economy can actually FIELD
// an army: at 50 metal/soldier a 4-6 soldier army (200-300 metal) drained most of a mine's
// output, so soldiers were never massed; at 30 metal/soldier that army costs 120-180 metal
// and is fundable alongside the (now cheaper) Outpost upkeep. The -200 money build cost and
// the -30 money salary are left UNCHANGED so soldiers stay a money commitment and pure-economy
// stays competitive. Mirrors soldier_cost() in rust-trainer/crates/cp-sim/src/resources.rs
// (parity-locked).
export const SOLDIER_COST: ResourceMap = rmap({
  [BasicResource.MONEY]: -200,
  [BasicResource.METAL]: -30,
});
export const SOLDIER_SALARY: ResourceMap = rmap({ [BasicResource.MONEY]: -30 });

File: rust-trainer/crates/cp-sim/src/resources.rs
//! Port of `src/core/resources.ts` (which itself ports `Course/basicresources.*`
//! and `Core/resourcemaps.h`).
//!
//! NUMERIC TYPE DECISION — resources are stored as `i64`.
//! The TS uses plain JS `number`s, but inspection of the model/managers shows
//! resource amounts are only ever produced by integer constants multiplied by
//! integer worker counts and combined with integer add/subtract (`merge`,
//! salaries, build costs). The only divisions in the codebase live in AI
//! heuristics and UI percentage bars — never in stored resource state. So
//! resources are integers in practice; `i64` is exact and avoids float drift,
//! which matters for deterministic parity with the TS golden traces.
//!
//! RESOURCEMAP SEMANTICS — `ResourceMap` mirrors the TS `Map<BasicResource,
//! number>` (which itself mirrors C++ `std::map<BasicResource,int>`). Several
//! algorithms (`merge_resource_maps`, `has_enough_resources`) rely on the fact
//! that iteration visits **only present keys**, and JS `Map` iterates in
//! insertion order. We therefore back the map with an insertion-ordered
//! `Vec<(BasicResource, i64)>` rather than a `HashMap`, so iteration order is
//! deterministic and matches the TS exactly.

/// `Course::BasicResource`. Discriminants match the TS enum verbatim.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum BasicResource {
    None = 0,
    Money = 1,
    Wood = 2,
    Stone = 3,
    Metal = 4,
}

/// Insertion-ordered map of resource -> amount.
///
/// Mirrors the TS `ResourceMap = Map<BasicResource, number>`: keys are unique,
/// iteration is in insertion order, and only present keys are visited.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct ResourceMap {
    entries: Vec<(BasicResource, i64)>,
}

impl ResourceMap {
    /// An empty map (TS `new Map()` / `EMPTY`).
    pub fn new() -> Self {
        ResourceMap { entries: Vec::new() }
    }

    /// Build from an ordered list of `(resource, amount)` pairs.
    ///
    /// Equivalent to the TS `rmap({...})` helper. Later duplicate keys
    /// overwrite earlier ones (matching `Map.set`), keeping the first
    /// position.
    pub fn from_pairs(pairs: &[(BasicResource, i64)]) -> Self {
        let mut m = ResourceMap::new();
        for &(k, v) in pairs {
            m.set(k, v);
        }
        m
    }

    /// `Map.get` — returns `None` when the key is absent.
    pub fn get(&self, key: BasicResource) -> Option<i64> {
        self.entries
            .iter()
            .find(|(k, _)| *k == key)
            .map(|(_, v)| *v)
    }

    /// `Map.has`.
    pub fn has(&self, key: BasicResource) -> bool {
        self.entries.iter().any(|(k, _)| *k == key)
    }

    /// `Map.set` — insert or overwrite, preserving insertion order for
    /// pre-existing keys.
    pub fn set(&mut self, key: BasicResource, value: i64) {
        if let Some(slot) = self.entries.iter_mut().find(|(k, _)| *k == key) {
            slot.1 = value;
        } else {
            self.entries.push((key, value));
        }
    }

    /// Number of present keys.
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// Whether the map has no keys.
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Iterate `(resource, amount)` pairs in insertion order.
    pub fn iter(&self) -> impl Iterator<Item = (BasicResource, i64)> + '_ {
        self.entries.iter().copied()
    }
}

/// `cloneResourceMap` — deep copy. (Rust `Clone` already does this; provided
/// for name-parity with the TS API.)
pub fn clone_resource_map(map: &ResourceMap) -> ResourceMap {
    map.clone()
}

/// `mergeResourceMaps`: start from `right`, then for every key in `left` either
/// insert it (if absent) or add to the existing value. Keys present in only one
/// map survive. (`Course::mergeResourceMaps`)
pub fn merge_resource_maps(left: &ResourceMap, right: &ResourceMap) -> ResourceMap {
    let mut result = right.clone();
    for (key, value) in left.iter() {
        match result.get(key) {
            None => result.set(key, value),
            Some(existing) => result.set(key, existing + value),
        }
    }
    result
}

/// `reverseResourceMap`: negate every value.
pub fn reverse_resource_map(map: &ResourceMap) -> ResourceMap {
    let mut result = ResourceMap::new();
    for (key, value) in map.iter() {
        result.set(key, value * -1);
    }
    result
}

/// `getNegativesMap`: keep negatives, zero out the rest.
pub fn get_negatives_map(map: &ResourceMap) -> ResourceMap {
    let mut result = ResourceMap::new();
    for (key, value) in map.iter() {
        result.set(key, if value < 0 { value } else { 0 });
    }
    result
}

/// `getPositivesMap`: keep `>= 0` values, zero out negatives.
pub fn get_positives_map(map: &ResourceMap) -> ResourceMap {
    let mut result = ResourceMap::new();
    for (key, value) in map.iter() {
        result.set(key, if value >= 0 { value } else { 0 });
    }
    result
}

// ---------------------------------------------------------------------------
// ConstResourceMaps and scalar economy constants (verbatim from resources.ts).
//
// The Mine / Hydroelectric / Nuclear values are a DELIBERATE rebalance away
// from the C++ original (see CLAUDE.md). The TS file is the source of truth for
// these three buildings — copied verbatim below.
// ---------------------------------------------------------------------------

use BasicResource::{Metal, Money, Stone, Wood};

/// Construct a const-style `ResourceMap` from pairs (helper for the builders
/// below; these are functions because `ResourceMap` owns a `Vec`).
fn rmap(pairs: &[(BasicResource, i64)]) -> ResourceMap {
    ResourceMap::from_pairs(pairs)
}

/// `EMPTY` — no keys at all.
pub fn empty() -> ResourceMap {
    ResourceMap::new()
}

/// `NO_RESOURCES` — all four resources present at 0 (insertion order: money,
/// wood, metal, stone — matching the TS literal order).
pub fn no_resources() -> ResourceMap {
    rmap(&[(Money, 0), (Wood, 0), (Metal, 0), (Stone, 0)])
}

/// `RESOURCE_LIMITS`.
pub fn resource_limits() -> ResourceMap {
    rmap(&[
        (Money, 9_999_999),
        (Wood, 9_999_999),
        (Stone, 9_999_999),
        (Metal, 9_999_999),
    ])
}

pub const UNIT_LIMITS: i64 = 999;

/// `STARTING_RESOURCES` — 400 money / 200 wood / 100 stone / 25 metal.
pub fn starting_resources() -> ResourceMap {
    rmap(&[(Money, 400), (Wood, 200), (Stone, 100), (Metal, 25)])
}

// Tile - Forest
pub fn forest_production() -> ResourceMap {
    rmap(&[(Wood, 100), (Stone, 10)])
}
pub fn forest_capacity() -> ResourceMap {
    rmap(&[(Wood, 600), (Stone, 60)])
}

// Tile - Abundant Forest
pub fn abundant_forest_production() -> ResourceMap {
    rmap(&[(Money, 15)])
}

pub const FOREST_GROW_TIME: i64 = 5;

// Building - Farm
pub fn farm_build_cost() -> ResourceMap {
    rmap(&[(Money, -100), (Wood, -100), (Metal, -5)])
}
pub fn farm_production() -> ResourceMap {
    rmap(&[(Money, 175)])
}
pub const FARM_GROW_TIME: i64 = 4;

// Building - Mine (REBALANCED — diverges from C++; TS is source of truth)
pub fn mine_build_cost() -> ResourceMap {
    rmap(&[(Money, -200), (Wood, -200), (Stone, 200)])
}
pub fn mine_production() -> ResourceMap {
    rmap(&[(Money, 20), (Stone, 30), (Metal, 20)])
}

// Building - Hydroelectric Power Plant (REBALANCED — TS is source of truth)
pub fn hepp_build_cost() -> ResourceMap {
    rmap(&[(Money, -280), (Wood, -150), (Stone, -120), (Metal, -60)])
}
pub fn hepp_production() -> ResourceMap {
    rmap(&[(Money, 80)])
}

// Building - Nuclear Power Plant (REBALANCED — TS is source of truth)
pub fn nuclearpp_build_cost() -> ResourceMap {
    rmap(&[(Money, -2000), (Wood, -200), (Stone, -250), (Metal, -250)])
}
pub fn nuclearpp_production() -> ResourceMap {
    rmap(&[(Money, 160)])
}

// Building - Outpost
pub fn outpost_build_cost() -> ResourceMap {
    rmap(&[(Money, -650), (Wood, -300), (Stone, -300), (Metal, -300)])
}
pub fn outpost_production() -> ResourceMap {
    rmap(&[(Money, -50), (Metal, -15)])
}
pub const OUTPOST_SOLDIER_VALUE: i64 = 3;

// Building - Bridge
pub fn bridge_build_cost() -> ResourceMap {
    rmap(&[(Money, -100), (Wood, -300), (Stone, -150)])
}
pub fn bridge_production() -> ResourceMap {
    rmap(&[(Wood, -5)])
}

// Building - Village (Neighborhood)
pub fn village_build_cost() -> ResourceMap {
    rmap(&[(Money, -200), (Wood, -200), (Stone, -100), (Metal, -25)])
}
pub fn village_production() -> ResourceMap {
    rmap(&[(Money, -10), (Wood, -10), (Stone, -10)])
}
pub const VILLAGE_UNIT_VALUE: i64 = 3;

// Building - Strange Device (DELIBERATE divergence — the original had no Device;
// see STRANGE-DEVICE-DESIGN.md / CLAUDE.md). One-time nuclear-tier build cost, NO
// per-turn drain; the balancer is the soldier-cap halving (see managers.rs
// update_unit_amounts). Mirrors STRANGE_DEVICE_* in src/core/resources.ts.
pub fn strange_device_build_cost() -> ResourceMap {
    // Tuned (2026-06-03): 1300 money (down from a 1800 nuclear-tier first pass). The
    // builder was 56% of bankruptcies at 1800 — 1300 keeps the Device a real commitment
    // while letting the (halved-army) builder stay solvent enough to DEFEND it, which
    // dropped hard-vs-hard bankruptcies 27%->~21%, raised the Device-win share to ~39%
    // of games, and put Device-survival near a coin-flip (~53%). Stays in lockstep with
    // STRANGE_DEVICE_BUILD_COST in src/core/resources.ts.
    rmap(&[(Money, -1300), (Stone, -200), (Metal, -200)])
}
pub const STRANGE_DEVICE_COUNTDOWN_BASE: i64 = 18;
pub const STRANGE_DEVICE_COUNTDOWN_PER_TILE: f64 = 0.12;
/// `strangeDeviceCountdown(tileCount)` — rounds before a standing Device wins.
/// `round(BASE + PER_TILE × tileCount)`. JS `Math.round` rounds half toward +∞;
/// Rust `f64::round` rounds half away from zero. They agree for the always-positive
/// values here (≥ 18), so `.round()` is the faithful port — do not "fix" it unless
/// BASE/PER_TILE ever go negative.
pub fn strange_device_countdown(tile_count: i64) -> i64 {
    (STRANGE_DEVICE_COUNTDOWN_BASE as f64 + STRANGE_DEVICE_COUNTDOWN_PER_TILE * tile_count as f64)
        .round() as i64
}

// Building - Mikontalo
pub const MIKONTALO_UNIT_VALUE: i64 = 2;

// Building - HQ
pub const HQ_UNIT_VALUE: i64 = 3;
pub const HQ_SOLDIER_VALUE: i64 = 1;

// Worker
pub fn basic_worker_cost() -> ResourceMap {
    rmap(&[(Money, -50)])
}
pub fn basic_worker_salary() -> ResourceMap {
    rmap(&[(Money, -5)])
}

// Expert
pub fn expert_cost() -> ResourceMap {
    rmap(&[(Money, -250)])
}
pub fn expert_salary() -> ResourceMap {
    rmap(&[(Money, -25)])
}

// Soldier
pub fn soldier_cost() -> ResourceMap {
    rmap(&[(Money, -200), (Metal, -50)])
}
pub fn soldier_salary() -> ResourceMap {
    rmap(&[(Money, -30)])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_get_has() {
        let mut m = ResourceMap::new();
        assert!(!m.has(Money));
        assert_eq!(m.get(Money), None);
        m.set(Money, 10);
        assert!(m.has(Money));
        assert_eq!(m.get(Money), Some(10));
        // overwrite keeps single entry
        m.set(Money, 20);
        assert_eq!(m.get(Money), Some(20));
        assert_eq!(m.len(), 1);
    }

    #[test]
    fn insertion_order_preserved() {
        let m = rmap(&[(Stone, 1), (Money, 2), (Metal, 3)]);
        let keys: Vec<BasicResource> = m.iter().map(|(k, _)| k).collect();
        assert_eq!(keys, vec![Stone, Money, Metal]);
    }

    #[test]
    fn merge_adds_shared_and_keeps_unique() {
        // right has money+wood, left has money+metal -> money sums, others kept
        let left = rmap(&[(Money, 100), (Metal, 5)]);
        let right = rmap(&[(Money, 1), (Wood, 7)]);
        let merged = merge_resource_maps(&left, &right);
        assert_eq!(merged.get(Money), Some(101));
        assert_eq!(merged.get(Wood), Some(7));
        assert_eq!(merged.get(Metal), Some(5));
        // result starts from `right`, so money & wood come first (right order),
        // then metal (new from left).
        let keys: Vec<BasicResource> = merged.iter().map(|(k, _)| k).collect();
        assert_eq!(keys, vec![Money, Wood, Metal]);
    }

    #[test]
    fn reverse_negates() {
        let m = rmap(&[(Money, 5), (Wood, -3)]);
        let r = reverse_resource_map(&m);
        assert_eq!(r.get(Money), Some(-5));
        assert_eq!(r.get(Wood), Some(3));
    }

    #[test]
    fn negatives_and_positives() {
        let m = rmap(&[(Money, 5), (Wood, -3), (Stone, 0)]);
        let neg = get_negatives_map(&m);
        assert_eq!(neg.get(Money), Some(0));
        assert_eq!(neg.get(Wood), Some(-3));
        assert_eq!(neg.get(Stone), Some(0));
        let pos = get_positives_map(&m);
        assert_eq!(pos.get(Money), Some(5));
        assert_eq!(pos.get(Wood), Some(0));
        // 0 counts as positive (>= 0)
        assert_eq!(pos.get(Stone), Some(0));
    }

    #[test]
    fn starting_resources_locked() {
        let s = starting_resources();
        assert_eq!(s.get(Money), Some(400));
        assert_eq!(s.get(Wood), Some(200));
        assert_eq!(s.get(Stone), Some(100));
        assert_eq!(s.get(Metal), Some(25));
    }

    #[test]
    fn rebalanced_industry_values_locked() {
        // These are the deliberate divergences from C++ — guard against drift.
        assert_eq!(nuclearpp_build_cost().get(Money), Some(-2000));
        assert_eq!(nuclearpp_production().get(Money), Some(160));
        assert_eq!(hepp_production().get(Money), Some(80));
        assert_eq!(mine_production().get(Metal), Some(20));
        assert_eq!(mine_build_cost().get(Stone), Some(200));
    }
}
